notes.md

```diff
@@ -0,0 +1,255 @@
+
+The PR shows how to implement steps 1 and 3 for `Array`s, `Diagonal`s, `Symmetric`s, and a custom `AbstractArray` `ScaledMatrix`.
+
+
+
+
+
+## Acceptable (Co)Tangents for `Array`
+
+Any `AbstractArray` is an acceptable (co)tangent for an `Array` (provided it's the right size, and its elements are appropriate (co)tangents for the elements of the primal `Array`).
+I'm going to assume this is true, because I can't see any obvious reason why it wouldn't be. 
+If anyone feels otherwise, please say.
+
+For example, this means that a `Diagonal{Float64}` is a valid (co)tangent for an `Array{Float64}`.
+
+
+
+
+
+## Optimising rrules for the equivalent function
+
+The basic example layed out above was very sub-optimal. Consider the following (equivalent) re-write
+```julia
+function rrule(config::RuleConfig, ::typeof(*), A::AbstractMatrix, B::AbstractMatrix)
+
+    # Produce pullbacks of destructure.
+    destructure_A_pb = pullback_of_destructure(config, A)
+    destructure_B_pb = pullback_of_destructure(config, B)
+
+    # Primal computation.
+    C = A * B
+
+    # Find pullback of restructure.
+    restructure_C_pb = pullback_of_restructure(config, C)
+
+    function my_mul_generic_pullback(C̄)
+
+        # Recover natural cotangent.
+        C̄_nat = restructure_C_pb(C̄)
+
+        # Compute pullback using natural cotangent of C.
+        Ā_nat = C̄_nat * B'
+        B̄_nat = A' * C̄_nat
+
+        # Transform natural cotangents w.r.t. A and B into structural (if non-primitive).
+        Ā = destructure_A_pb(Ā_nat)
+        B̄ = destructure_B_pb(B̄_nat)
+        return NoTangent(), Ā, B̄
+    end
+
+    return C, my_mul_generic_pullback
+end
+```
+A few observations:
+1. All dense primals are gone. In the pullback, they only appeared in places where they can be safely replaced with the primals themselves because they're doing array-like things. `C_dense` appeared in the construction of `restructure_C_pb`, however, we were using a sub-optimal implementation of that function. Much of the time, `restructure_of_pb` doesn't require `C_dense` in order to know what the pullback would look like and, if it does, it can be obtained from `C`.
+2. All direct calls to `rrule_via_ad` have been replaced with calls to functions which are defined to return the things we actually need (the pullbacks). These typically have efficient (and easy to write) implementations.
+3. `C̄_nat` could be any old `AbstractArray`, because it's conceptually the cotangent for an `Array`. For example, `pullback_of_restructure` for a `Diagonal` returns a `Diagonal`. This is good -- it means we might occassionally get faster computations in the pullback. 
```

Comment:
Agreed. I didn't phrase this ideally.

I'm trying to say that it's okay for the `pullback_of_restructure` to return a `Diagonal` in this context. This won't happen if we just use AD though -- we would need to manually arrange for this to happen.

When I wrote this, I think I had already implemented something to make sure that this happens.

Reply:
How about
```suggestion
3. It is permissible for `C̄_nat` to be any old `AbstractArray`, because it's conceptually the cotangent for an `Array`. For example, it is permissible for us to manually optimise `pullback_of_restructure` for a `Diagonal` to ensure that it returns a `Diagonal`. This is good -- it means we can occassionally get faster computations in the natural pullback. 
```
?